msprime/stats.py

```diff
@@ -711,3 +720,155 @@ def wfn(x):
+        if windows[-1] != self.tree_sequence.sequence_length:
+            raise ValueError("Windows must extend to the end of the sequence.")
+        for k in range(num_windows):
+            if windows[k + 1] <= windows[k]:
+                raise ValueError("Windows must be increasing.")
+        num_sample_sets = len(sample_sets)
+        n = [len(x) for x in sample_sets]
+        n_out = len(weight_fun([0 for a in range(num_sample_sets)]))
+
+        # get the mutation stuff
+        sites = tables.SiteTable()
+        mutations = tables.MutationTable()
+        self.tree_sequence.dump_tables(sites=sites, mutations=mutations)
+        nsites = sites.num_rows
+        nmuts = mutations.num_rows
+
+        # we store the final answers here
+        S = [[0.0 for j in range(n_out)] for _ in range(num_windows)]
+        if nmuts == 0:
+            return S
+        # print("sample_sets:", sample_sets)
+        # print("n_out:",n_out)
+        N = self.tree_sequence.num_nodes
+        # initialize: with no tree, each node is either in a sample set or not
+        X = [[int(u in a) for a in sample_sets] for u in range(N)]
+        # we will construct the tree here
+        pi = [-1 for j in range(N)]
+        # keep track of where we are for the windows
+        chrom_pos = 0.0
+        # keep track of which site we're looking at
+        site_num = 0
+        next_mut = 0
+        assert mutations.site[0] == 0
+        # index of *left-hand* end of the current window
+        window_num = 0
+        for interval, records_out, records_in in self.tree_sequence.edge_diffs():
+            if site_num == nsites:
+                break
+            for sign, records in ((-1, records_out), (+1, records_in)):
+                for edge in records:
+                    # print("Record (",sign,"):",node,children,time)
+                    # print("\t",X, "-->", L)
+                    dx = [0 for k in range(num_sample_sets)]
+                    if sign == +1:
+                        pi[edge.child] = edge.parent
+                    for k in range(num_sample_sets):
+                        dx[k] += sign * X[edge.child][k]
+                    if sign == -1:
+                        pi[edge.child] = -1
+                    for k in range(num_sample_sets):
+                        X[edge.parent][k] += dx[k]
+                    # propagate change up the tree
+                    u = pi[edge.parent]
+                    if u != -1:
+                        next_u = pi[u]
+                        while u != -1:
+                            for k in range(num_sample_sets):
+                                X[u][k] += dx[k]
+                            u = next_u
+                            next_u = pi[next_u]
+                    # print("\t",X, "-->", L)
+            # print("next tree:",L,length)
+            site_pos = sites.position[site_num]
+            start_mut = next_mut
+            assert mutations.site[start_mut] == site_num
```

Comment:
Here I'm assuming that every site has some mutations.  Is this something we assume, @jeromekelleher?  It'd be easy to not assume it if  not already.

Reply:
We can have sites that have zero mutations; this turns out to be useful for me when doing inference.